Firestore/Source/Remote/FSTStream.mm

```diff
@@ -525,11 +525,7 @@ - (void)handleStreamMessage:(id)value {
  */
 - (void)handleStreamClose:(nullable NSError *)error {
   FSTLog(@"%@ %p close: %@", NSStringFromClass([self class]), (__bridge void *)self, error);
-
-  if (![self isStarted]) {  // The stream could have already been closed by the idle close timer.
-    FSTLog(@"%@ Ignoring server close for already closed stream.", NSStringFromClass([self class]));
-    return;
-  }
+  FSTAssert([self isStarted], @"handleStreamClose: called for non-started stream.");
```

Comment:
I don't think this is safe at all.

PR #811 was defending against gRPC calling us and enqueueing an error while (or just before) we were already processing a close for a different reason. This pushed the check in the callback filter out onto the dispatch queue making the callback filter effective again and prevents callbacks from gRPC from getting us after the stream is closed.

However, it's still the case that an idle close timer firing, a user disableNetwork, and a gRPC error can be enqueued at the same time. The first one to succeed may now prevent gRPC from kicking out any further events, but the subsequent close actions are still going to pass through here.

Reply:
Agree to disagree.

Whichever of those 3 are executed first will call `closeWithFinalState:` which will call `suppressCallbacks` in conjunction with setting `state` to a non-started value, which will prevent subsequent calls to `handleStreamClose:`.

The only other caller of `handleStreamClose:` `is resumeStartWithToken:` when handling auth failures, but it asserts that `self.state == FSTStreamStateAuth` immediately prior to calling `handleStreamClose:` so this assert is safe in that case too.

Reply:
Said differently, we always suppress callbacks when setting the stream state to a non started state...  And with your change we can now rely on that to prevent any subsequent handleStreamClose calls.

Reply:
Okay, I've convinced myself you're right by tracing it through. All callers check the state prior to calling so this is safe.